### PR TITLE
MiniRunners + OperatingUnits

### PR DESCRIPTION
**Pull request by 17zhangw**
This PR contains the following:
- mini_runners (in `benchmark/runner/mini_runners`)
- Addition of `key_size` and `num_keys` to feature
- Removed `NLJOIN_LEFT/RIGHT` as output features
- Fix `PlanGenerator` generation of HashJoinPlanNode
- Removed useless keys from NestedLoopJoinPlanNode

### CODE REVIEW

File: benchmark/runner/mini_runners.cpp [hunk 1]
```diff
@@ -0,0 +1,1161 @@
...
+    brain::ExecutionOperatingUnitFeatureVector pipe0_vec;
+    brain::ExecutionOperatingUnitFeatureVector pipe1_vec;
+    pipe0_vec.emplace_back(brain::ExecutionOperatingUnitType::SEQ_SCAN, row, 4 * num_col, num_col,
+                           static_cast<double>(car));
+    pipe0_vec.emplace_back(brain::ExecutionOperatingUnitType::HASHJOIN_BUILD, row, 4 * num_col, num_col,
+                           static_cast<double>(car));
+    pipe1_vec.emplace_back(brain::ExecutionOperatingUnitType::SEQ_SCAN, row, 4 * num_col, num_col,
+                           static_cast<double>(car));
+    pipe1_vec.emplace_back(brain::ExecutionOperatingUnitType::HASHJOIN_PROBE, row, 4 * num_col, num_col,
+                           static_cast<double>(car));
```
Comment (malin1993ml) [inline benchmark/runner/mini_runners.cpp]: Unlike other operating units in the execution engine, the `car` here should be the number of rows matched by the hash table look up from the probe table. You can refer [here](https://github.com/malin1993ml/terrier/blob/modeling-metrics/script/model/execution_engine_runner/generate_hashjoin.py#L185) on how I get this number. Note that in that script I hacked the hash join key comparison function so that the hash join only uses the value of the last column. You might need to change all the column `Dist` types to `Rotate` in the `TableGenerator`, and make every column to have the same value (maybe by fixing the random seed)?

You may also want to limit the total cardinality here by controlling the input argument for this runner, like what I did [here](https://github.com/malin1993ml/terrier/blob/modeling-metrics/script/model/execution_engine_runner/generate_hashjoin.py#L189). Otherwise the matched tuple can explode. Similar to the `NLJoin`. Also be careful there because every pair of tuples are going to be "matched" in a nested loop join.

We can meet to discuss this if it's unclear.

File: benchmark/runner/mini_runners.cpp [hunk 2]
```diff
@@ -0,0 +1,1161 @@
...
+      UNREACHABLE("Unexpected number of columns for UpdateRunners");
+      break;
+  }
+
+  // NOLINTNEXTLINE
+  for (auto _ : state) {
+    // Create temporary table schema
+    std::stringstream query;
+    query << "UPDATE "
+          << "INTEGERCol15Row" << num_rows << "Car1 SET ";
```
Comment (malin1993ml) [inline benchmark/runner/mini_runners.cpp]: Extract a table name generation static function in `TableGenerator` that takes in the type, the row number, the column number, and the cardinality. And use the same function both in the `TableGenerator` and all the table name references here.

File: benchmark/runner/mini_runners.cpp [hunk 3]
```diff
@@ -0,0 +1,1161 @@
...
+
+  static constexpr execution::query_id_t NLJ_COL_1_QID = execution::query_id_t(33);
+  static constexpr execution::query_id_t NLJ_COL_4_QID = execution::query_id_t(34);
+  static constexpr execution::query_id_t NLJ_COL_8_QID = execution::query_id_t(35);
+  static constexpr execution::query_id_t NLJ_COL_15_QID = execution::query_id_t(36);
+
+  static constexpr execution::query_id_t IDX_KEY_1_QID = execution::query_id_t(37);
+  static constexpr execution::query_id_t IDX_KEY_4_QID = execution::query_id_t(38);
+  static constexpr execution::query_id_t IDX_KEY_8_QID = execution::query_id_t(39);
+  static constexpr execution::query_id_t IDX_KEY_15_QID = execution::query_id_t(40);
```
Comment (malin1993ml) [inline benchmark/runner/mini_runners.cpp]: I don't think we really want to do this. I think we can just use a simple counter and increment with every query we run. Otherwise we have to fix these numbers every time want to add/modify the number of columns we're going to test for a operating unit.

File: benchmark/runner/mini_runners.cpp [hunk 4]
```diff
@@ -0,0 +1,1161 @@
...
+ * 2 - Cardinality
+ */
+static void GenScanArguments(benchmark::internal::Benchmark *b) {
+  auto num_cols = {1, 4, 8, 15};
+  auto num_rows = {
+      std::pair<int, int>{10, 1},           std::pair<int, int>{10, 2},          std::pair<int, int>{10, 8},
+      std::pair<int, int>{100, 16},         std::pair<int, int>{100, 32},        std::pair<int, int>{100, 64},
+      std::pair<int, int>{10000, 128},      std::pair<int, int>{10000, 1024},    std::pair<int, int>{10000, 4096},
+      std::pair<int, int>{10000, 8192},     std::pair<int, int>{1000000, 1024},  std::pair<int, int>{1000000, 16384},
+      std::pair<int, int>{1000000, 131072}, std::pair<int, int>{1000000, 524288}};
```
Comment (malin1993ml) [inline benchmark/runner/mini_runners.cpp]: I think we can just generate an exponentially increased array for the cardinality based on the row num. Just similar to what I did [here](https://github.com/malin1993ml/terrier/blob/modeling-metrics/script/model/execution_engine_runner/generate_scan.py#L61). I think manually maintaining these pairs can be cumbersome if we want to boost the amount of training data later.

File: benchmark/runner/mini_runners.cpp [hunk 5]
```diff
@@ -0,0 +1,1161 @@
...
+ * 1 - Iteration count
+ */
+static void GenArithArguments(benchmark::internal::Benchmark *b) {
+  auto operators = {brain::ExecutionOperatingUnitType::OP_INTEGER_PLUS_OR_MINUS,
+                    brain::ExecutionOperatingUnitType::OP_INTEGER_MULTIPLY,
+                    brain::ExecutionOperatingUnitType::OP_INTEGER_DIVIDE,
+                    brain::ExecutionOperatingUnitType::OP_INTEGER_COMPARE,
+                    brain::ExecutionOperatingUnitType::OP_DECIMAL_PLUS_OR_MINUS,
+                    brain::ExecutionOperatingUnitType::OP_DECIMAL_MULTIPLY,
+                    brain::ExecutionOperatingUnitType::OP_DECIMAL_DIVIDE};
```
[the remark this answers is not in the document]
Comment (17zhangw) [reply benchmark/runner/mini_runners.cpp]: Added.

File: benchmark/runner/mini_runners.cpp [hunk 6]
```diff
@@ -0,0 +1,1161 @@
...
+    return iterator;                                               \
+  }
+
+#elif __GNUC__
+#define TIGHT_LOOP_OPERATION(type, name, op)                                \
+  __attribute__((noinline, noclone)) type __##type##_##name(size_t count) { \
+    type iterator = 1;                                                      \
+    for (size_t i = 1; i <= count; i++) {                                   \
+      iterator = iterator op i;                                             \
+      DoNotOptimizeAway(i);                                                 \
```
[the remark this answers is not in the document]
Comment (17zhangw) [reply benchmark/runner/mini_runners.cpp]: Protection on `i` is not actually necessary. Has been removed.

File: benchmark/runner/mini_runners.cpp [hunk 7]
```diff
@@ -0,0 +1,1161 @@
...
+
+      auto query_info = optimizer::QueryInfo(parser::StatementType::SELECT, std::move(output), &property_set);
+      out_plan =
+          optimizer.BuildPlanTree(txn, accessor.get(), db_main->GetStatsStorage().Get(), query_info, std::move(plan));
+    } else {
+      auto property_set = optimizer::PropertySet();
+      auto query_info = optimizer::QueryInfo(stmt_list->GetStatement(0)->GetType(), {}, &property_set);
+      out_plan =
+          optimizer.BuildPlanTree(txn, accessor.get(), db_main->GetStatsStorage().Get(), query_info, std::move(plan));
+    }
```
[the remark this answers is not in the document]
Comment (17zhangw) [reply benchmark/runner/mini_runners.cpp]: Swapped for `trafficcop::TrafficCopUtil::Optimize()`.

File: benchmark/runner/mini_runners.cpp (hunk 3 above)
Comment (17zhangw) [reply benchmark/runner/mini_runners.cpp]: Done.

File: src/include/planner/plannodes/nested_loop_join_plan_node.h [hunk 8]
```diff
@@ -119,17 +72,6 @@ class NestedLoopJoinPlanNode : public AbstractJoinPlanNode {
...
- private:
-  /**
-   * Left keys (derived from child 0) to join on
-   */
-  std::vector<common::ManagedPointer<parser::AbstractExpression>> left_keys_;
-
-  /**
-   * Right keys (derived from child 1) to join on
-   */
-  std::vector<common::ManagedPointer<parser::AbstractExpression>> right_keys_;
```
[the remark this answers is not in the document]
Comment (17zhangw) [reply src/include/planner/plannodes/nested_loop_join_plan_node.h]: Yes, these are intended changes.

File: src/brain/operating_unit_recorder.cpp [hunk 9]
```diff
@@ -36,135 +37,125 @@
 
 namespace terrier::brain {
 
-type::TypeId OperatingUnitRecorder::DeriveComputation(common::ManagedPointer<parser::AbstractExpression> expr) {
-  auto lchild = expr->GetChild(0);
-  if (lchild->GetReturnValueType() != type::TypeId::INVALID &&
-      lchild->GetReturnValueType() != type::TypeId::PARAMETER_OFFSET) {
-    return lchild->GetReturnValueType();
+size_t OperatingUnitRecorder::ComputeKeySize(
+    const planner::AbstractPlanNode *plan,
```
[the remark this answers is not in the document]
Comment (17zhangw) [reply src/brain/operating_unit_recorder.cpp]: Refactored.

File: src/brain/operating_unit_recorder.cpp [hunk 10]
```diff
@@ -36,135 +37,125 @@
...
-        case type::TypeId::DATE:
-          return ExecutionOperatingUnitType::OP_INTEGER_COMPARE;
-        case type::TypeId::VARCHAR:
-        case type::TypeId::VARBINARY:
-          // TODO(wz2): Revisit this since VARCHAR/VARBINARY is more than just integer
-          return ExecutionOperatingUnitType::OP_INTEGER_COMPARE;
-        default:
-          return ExecutionOperatingUnitType::INVALID;
-      }
+size_t OperatingUnitRecorder::ComputeKeySize(const planner::AbstractPlanNode *plan, catalog::table_oid_t tbl_oid) {
```
[the remark this answers is not in the document]
Comment (17zhangw) [reply src/brain/operating_unit_recorder.cpp]: Refactored.

File: src/brain/operating_unit_recorder.cpp [hunk 11]
```diff
@@ -36,135 +37,125 @@
...
+      return;
     }
+  }
+
+  pipeline_features_.emplace(type, ExecutionOperatingUnitFeature(type, num_rows, key_size, num_keys, cardinality));
+}
 
-    for (auto child : head->GetChildren()) {
-      work.push(child);
+void OperatingUnitRecorder::RecordFeatures(const planner::AbstractPlanNode *plan, size_t scaling) {
```
[the remark this answers is not in the document]
Comment (17zhangw) [reply src/brain/operating_unit_recorder.cpp]: Refactored.

File: test/execution/compiler_test.cpp [hunk 12]
```diff
@@ -2075,10 +2077,10 @@ TEST_F(CompilerTest, SimpleNestedLoopJoinTest) {
...
   // NLJOIN left and right are in same pipeline
+  // But NLJOIN left/right features do not exist
   auto feature_vec0 = pipeline->GetPipelineFeatures(execution::pipeline_id_t(0));
   auto exp_vec0 = std::vector<brain::ExecutionOperatingUnitType>{
-      brain::ExecutionOperatingUnitType::NLJOIN_LEFT, brain::ExecutionOperatingUnitType::NLJOIN_RIGHT,
-      brain::ExecutionOperatingUnitType::OP_INTEGER_COMPARE,
+      brain::ExecutionOperatingUnitType::OP_INTEGER_COMPARE, brain::ExecutionOperatingUnitType::OP_INTEGER_COMPARE,
       brain::ExecutionOperatingUnitType::OP_INTEGER_PLUS_OR_MINUS, brain::ExecutionOperatingUnitType::SEQ_SCAN};
   EXPECT_TRUE(CheckFeatureVectorEquality(feature_vec0, exp_vec0));
 }
```
[the remark this answers is not in the document]
Comment (17zhangw) [reply test/execution/compiler_test.cpp]: Yes.

File: src/brain/operating_unit_recorder.cpp [hunk 13]
```diff
@@ -174,11 +165,64 @@ void OperatingUnitRecorder::VisitAbstractPlanNode(const planner::AbstractPlanNod
...
+  auto features = OperatingUnitUtil::ExtractFeaturesFromExpression(plan->GetScanPredicate());
   plan_features_.insert(plan_features_.end(), std::make_move_iterator(features.begin()),
                         std::make_move_iterator(features.end()));
 }
 
+void OperatingUnitRecorder::Visit(const planner::SeqScanPlanNode *plan) {
+  VisitAbstractScanPlanNode(plan);
+  RecordFeatures(plan, 1);
+
+  // For a sequential scan, # keys is the number of columns output
```
[the remark this answers is not in the document]
Comment (17zhangw) [reply src/brain/operating_unit_recorder.cpp]: I think the comment might be misleading. In reality, what ends up being recorded is the number of columns being scanned. I have updated the in-code comment.

File: benchmark/runner/mini_runners.cpp (hunk 1 above)
Comment (17zhangw) [reply benchmark/runner/mini_runners.cpp]: Changed this such that `col15` is used multiple times.

File: benchmark/runner/mini_runners.cpp (hunk 4 above)
Comment (17zhangw) [reply benchmark/runner/mini_runners.cpp]: Changed.

File: util/execution/table_generator/table_generator.cpp [hunk 14]
```diff
@@ -271,6 +271,71 @@ void TableGenerator::GenerateTestTables(bool is_mini_runner) {
...
+        auto index_oid = exec_ctx_->GetAccessor()->CreateIndex(ns_oid_, table_oid, idx_name.str(), tmp_index_schema);
+        auto &index_schema = exec_ctx_->GetAccessor()->GetIndexSchema(index_oid);
+        index_builder.SetKeySchema(index_schema);
+        auto *tmp_index = index_builder.Build();
+        exec_ctx_->GetAccessor()->SetIndexPointer(index_oid, tmp_index);
+        auto index = exec_ctx_->GetAccessor()->GetIndex(index_oid);
+
+        // Fill up the index
+        auto idx_meta = IndexInsertMeta("", "", std::move(idx_meta_cols));
+        FillIndex(index, index_schema, idx_meta, table, schema);
```
[the remark this answers is not in the document]
Comment (17zhangw) [reply util/execution/table_generator/table_generator.cpp]: Extracted and simplified logic.

File: benchmark/runner/mini_runners.cpp [hunk 15]
```diff
@@ -0,0 +1,889 @@
...
+                           row * row / car);
+    pipe1_vec.emplace_back(brain::ExecutionOperatingUnitType::OP_INTEGER_COMPARE, row, 4, 1, row * row / car);
+    units.RecordOperatingUnit(execution::pipeline_id_t(0), std::move(pipe0_vec));
+    units.RecordOperatingUnit(execution::pipeline_id_t(1), std::move(pipe1_vec));
+
+    std::stringstream query;
+    auto tbl_name = execution::sql::TableGenerator::GenerateTableName(type::TypeId::INTEGER, 15, row, car);
+    query << "SELECT b.col15 FROM " << tbl_name << ", " << tbl_name << " as b WHERE ";
+    for (auto i = 1; i <= num_col; i++) {
+      query << tbl_name << ".col15 = b.col15";
```
Comment (malin1993ml) [inline benchmark/runner/mini_runners.cpp]: Ah! I was actually thinking about duplicating the columns when generating the tables in `table_generator.cpp`. Is that doable/easy to do? Because I'm a bit concerned that the optimizer may simplify the duplicated join predicate and we only end up with joining on one column. Even though our optimize may not optimize this now, later modifications (like the new rewriting rules from Erik) may just change the behavior.

Comment (17zhangw) [reply benchmark/runner/mini_runners.cpp]: Tweaked. `table_generator.cpp` now generates col16-col31 which duplicate the values of col15. The join predicate for hash join and nested loop join would then use col15 through col31.
